### ISSUE TITLE
[BUG] Choppy frame rate, makes game unplayable

### ISSUE THREAD

**Issue by owner-w** [atsb/Doom64EX-Plus#198]
I am getting a lot of frame skips, leading to a very choppy "low fps" experience. The stutter may seem tolerable at first but it is really hard to aim at monsters during intense battles, making it difficult to play. Video demonstration:

https://user-images.githubusercontent.com/132102644/235251848-4b07ec7d-744a-4ffc-95a0-580256c0f811.mp4

My resolution is 3840×2160@120hz.

Regular EX does not have this issue. See video of how it behaves there:

https://user-images.githubusercontent.com/132102644/235251931-4206baba-ac52-4f33-8f16-cf176597af2e.mp4

**Comment by atsb** [comment]
This is Windows or a Linux distro?  Because I do get some frame skips on Linux and I’ve been attempting to track this down for months.  It only happens on uncapped mode too.

**Comment by atsb** [comment]
I've narrowed it down to at least (what I think) is a combination of vsync and hardware.  On one of my 'not so good' machines running Linux, it is dradful and very choppy.  On Windows, I don't see it at all, it is very smooth and on my powerful Linux machine with a 12 core Xeon and RTX A2000 it very rarely skips a frame.

Try without vsync and see if it helps.

**Comment by azamorapl** [comment]
For what it's worth I had a similar issue, and I fixed it by disabling vsync in Nvidia Control Panel (the built in options did nothing).

**Comment by bubbleguuum** [comment]
EX+ has weird frame pacing issues depending on settings.
On my 4K 60hz G-Sync monitor I had to do all of these to have no stutter (micro or not), no (massive) input lag and no tearing:

- In the NVIDIA CP settings for doom64ex+.exe, set 'Monitor technology' to 'Fixed refresh', to disable G-Sync
- in the game, set V-Sync to Off
- use RTSS framerate limiter to limit framerate to 60 (use your monitor max refresh rate here) for doom64ex+.exe

**Comment by atsb** [comment]
Yeah, I thought I fixed it but I got it on one of my machines.  This is an issue I've been trying to figure out.  The original EX was also plagued ever so slightly by the same thing but somewhere, frame pacing is being delayed by what seems like 1ms every few frames.

**Comment by atsb** [comment]
Ok, I have released RC2
https://github.com/atsb/Doom64EX-Plus/releases/download/3.6.5.8-rc1/doom64ex-plus-3.6.5.8-RC2-windows.zip

I tested this on a rather low-end compuer with an i3-1115G4 (UHD Graphics G4) machine.  I had zero stutters and it was buttery smooth (Windows 11).

**Comment by bubbleguuum** [comment]
Thanks I did some test with that version, GTX 1080TI, 4K, fullscreen mode, in-game vsync off, no NVIDIA CP tweaks, no RTSS framerate cap.

All stutter seems to be indeed gone. But there is still some input lag (keyboard and mouse).

Using RTSS (Riva Tuner Server Statistics, part of MSI Afterburner)  to cap framerate to 60 (my monitor's refresh rate) entirely fixes input lag, making input super responsive.
RTSS with the framerate cap produces super stable frametimes which are exactly 1000/60 = 16.6667 ms. This can be seen with its framegraph display.

Also I'm not sure the in-game "vsync off" setting works properly as (without the RTSS framecap) I got no screen tearing at all as if it was vsync on.

**Comment by atsb** [comment]
Thanks for this, good news indeed.  I'll clean up the latency and get the vsync working.

**Comment by atsb** [comment]
Ok, feels like I have removed latency on movement now and I placed the vsync toggle within the code that draws frames, so each frame it'll be set instead of when just the renderer was active.

I also fixed the Interpolation setting crashing when used a lot and fixed the audio (my little experimental hack didn't work).  I re-uploaded a new RC2.

**Comment by azamorapl** [comment]
Tried with RC2, I still have the same stutter as before. I am using a 120hz monitor, though.

**Comment by atsb** [comment]
Yeah, I can’t test that at all.  I might just cap the frame rate to 60 and no higher

**Comment by bubbleguuum** [comment]
> Tried with RC2, I still have the same stutter as before. I am using a 120hz monitor, though.

With v-sync off in-game ?

**Comment by atsb** [comment]
I’ll do some experiments on my more powerful laptop later without any vsync implementation.  Perhaps this could be causing some issues at higher refresh rates.  The EX code is pretty weird in places and it maybe is interfering with the interpolation.

**Comment by azamorapl** [comment]
> > Tried with RC2, I still have the same stutter as before. I am using a 120hz monitor, though.
> 
> With v-sync off in-game ?

Yeah, it makes no difference (or maybe slightly better? I can't tell if it's a placebo). Still pretty much like the video in the OP. Everything is fine if I explicitly disable vsync for the game in nvidia control panel, though.

**Comment by bubbleguuum** [comment]
> > > Tried with RC2, I still have the same stutter as before. I am using a 120hz monitor, though.
> > 
> > 
> > With v-sync off in-game ?
> 
> Yeah, it makes no difference. Plays pretty much like the video in the OP. Everything is fine if I explicitly disable vsync for the game in nvidia control panel, though.

Interesting, as in-game vsync off in RC2 really behaves as if it was on, according to my testing.

**Comment by atsb** [comment]
Ok, I got it.  I tried on my Linux machine which has an RTX GPU and yeah I still got stuttering, unless vsync was off in NVidia control panel.  I have removed VSync and Antialiasing from the code and now, I get a totally smooth frametime graph on Linux (and probably Windows too).  I'll do a little more fine tuning and push the code

**Comment by azamorapl** [comment]
Tried the latest build but the stuttering is still there for me, probably due to the higher refresh rate?

**Comment by atsb** [comment]
Must be, I don’t get it at all on any of my systems

**Comment by atsb** [comment]
I'll look into getting it working with higher refresh rates, but since this is now fixed for the majority of users, I'll close it.

**Comment by bubbleguuum** [comment]
There is still framepacing issues and input lag in released v3.6.5.8. 
Tested on Windows 11, 60Hz 4K monitor with 4K resolution set in-game (using Windows scaling workaround), G-Sync disabled, no NVIDIA CP tweaks, first level on WMD difficulty.

And there is noticeable stutter, but only in fullscreen mode, as shown by these framegraph spikes (fullscreen, no frame cap):

![fullscreen_noframecap](https://github.com/atsb/Doom64EX-Plus/assets/761164/2622d69d-1706-4528-9b25-49c2ffe62a93)

In windowed mode, there is no lag spikes (as long as `v_windowborderless` is not enabled (set to 1)) but there is still noticeable input lag (windowed, no framecap):

![windowed_noframecap](https://github.com/atsb/Doom64EX-Plus/assets/761164/7a84e4ce-0123-4120-b9d5-b2af39cbfeb7)

To get rid of input lag and have perfect 16.666ms frametimes most of the time, I still have to cap framerate to 60 with RTSS and use windowed mode (windowed, frame cap):

![windowed_framecap](https://github.com/atsb/Doom64EX-Plus/assets/761164/16b07224-e87d-4711-9faf-3689339f3b45)

So long story short: it works best in windowed mode with a 60Hz RTSS framecap. Forcing vsync in NVIDIA CP does nothing for me.

**Comment by atsb** [comment]
It must need to be capped to 60hz then internally, I’ll make sure to add this so the frame rate will have this internal cap.  Not nice, but a solution.  Thanks!

**Comment by ghost** [comment]
Sorry to necro this, but I'm getting _horrible_ framerates playing this in Linux right now via the flatpak.  It's labeled version 4.0.0.3.SDL.3.2.10  I've turned off vsync in the game and tried turning off adaptive sync just to be sure (though adaptive sync should have no issue with 60 FPS if that's what this game wants.)  I don't have a control panel (AMD GPU) but the amdgpu driver does not force or default to vsync on (part of why I want adaptive sync on.)  I'm using Debian Trixie (13) and KDE Plasma on Wayland, so I wonder if it just works better with X11 or something.

It really feels like the framerate is around 10 FPS or so.  (Yeah, not even 15.)  Interpolation helps, but it's still making me feel sick (I need smooth motion or I get motion sickness.)  I've tried vsync on and off, windowed on and off, all with the same result.  Normally I keep adaptive sync set to automatic which means it usually only triggers on in games as far as I know, but I also tried setting it to never and always with no difference that I could tell in this.  This is on a 144Hz monitor.

CPU usage and GPU usage go up maybe 2% when playing the game.  Definitely no limitations there.  (9900X CPU, 7800XT GPU, IGP disabled, so there should be no problem here.)  I no longer have a Windows system to test in (since MS has lost their everloving minds, I didn't bother buying a license for this system.)  It crashes in WINE though, so I can't test that way.

**Comment by atsb** [comment]
Does this happen on your system if compiling from source?  Just curious, I don't think it has anything to do with flatpak.

**Comment by ghost** [comment]
Looks like compiling from source produces the same result.

EDIT:  I just tried X11 instead of Wayland with the same result.
